### ISSUE TITLE
Edit card with API ?

### ISSUE THREAD

**Issue by FabianSeidel** [wekan/wekan#3283]
Hi,

I am trying to edit cards via REST PUT but I get the error Access-Control-Allow-Methods: Method PUT is not allowed by Access-Control-Allow-Methods.

Running Wekan on SNAP and GET POST works fine. 

BR Fabian

**Comment by FabianSeidel** [comment]
Any idea? Its definetly a CORS problem because when I block CORS in chrome it works. 

The only way I do see to solve the problem is to start a new setup based in docker, does this solve the Problem?

**Comment by xet7** [comment]
https://github.com/wekan/wekan/blob/master/docker-compose.yml#L291-L299

For snap, lowercase and underscore to minun, for example:
```
sudo snap set cors=`*`
```
Removing setting:
```
sudo snap unset cors
```
Other settings at Wekan server bash:
```
wekan.help | less
```

**Comment by FabianSeidel** [comment]
CORS did already work but still the issue with PUT / axios. I set the following commands but still can't edit any cards:

sudo snap set wekan cors='https://myURL' 
sudo snap set wekan cors-allow-headers='Authorization,Content-Type' 
sudo snap set wekan cors-expose-headers='*'

**Comment by FabianSeidel** [comment]
Maybe smith wrong with my api call?

await this.$axios({
        url: `myURL/api/boards/${cardSave.boardId}/lists/${cardSave.listId}/cards/${cardSave._id}?access_token=${this.$state.token.token}`,
        method: 'put',
        timeout: 0,
        headers: {
          'Content-Type': 'application/json'
        },
        data: JSON.stringify(data)
      })

=> Get and POST do work but PUT still give me the error: Method PUT is not allowed by Access-Control-Allow-Methods.

**Comment by xet7** [comment]
@FabianSeidel 

Are you using admin username and password to login to API and get bearer token?

**Comment by FabianSeidel** [comment]
Yes, I use admin account to get the token. When I block CORS policy PUT works :/

**Comment by xet7** [comment]
Did you just unset cors settings? Then it did work?

**Comment by xet7** [comment]
How did you block it ?

**Comment by FabianSeidel** [comment]
With Chrome browser extension, not on server site

**Comment by FabianSeidel** [comment]
I guess something like this has to be added: Access-Control-Allow-Methods: POST, GET => I may try a setup with docker and add this to the docker file manually

**Comment by ItamarSoffer** [comment]
Hi, is there an update for the issue? I can't edit cards and move them between lists either.

**Comment by xet7** [comment]
@ItamarSoffer 

If you mean drag-drop: According to https://github.com/wekan/wekan/issues/3647#issuecomment-808007007 edit by moving and move with drag-drop works.

If you mean API: There is https://github.com/wekan/wekan/pull/3672 that fixes generating API docs, it will be in next release, then all of API is visible again at https://wekan.github.io/api/

Edit REST API code is here https://github.com/wekan/wekan/blob/master/models/cards.js#L2567

**Comment by ItamarSoffer** [comment]
I meant moving a card with the API (using the `PUT /api/boards/{board}/lists/{list}/cards/{card}` function.
When I send the request I get a 403 Error.

**Comment by xet7** [comment]
@ItamarSoffer 

Did you login to API first to get bearer token? Please see Python example about how to add card:
https://github.com/wekan/wekan/blob/master/api.py

**Comment by ItamarSoffer** [comment]
Of course. 
The POST and GET works great (I’m able to create a new card with the API), but the PUT and DELETE won’t. 
( I’m the admin of the specific board)

**Comment by xet7** [comment]
@ItamarSoffer 

You can spend some quality time looking at code somewhere in this file and send pull request to fix it:
https://github.com/wekan/wekan/blob/master/models/cards.js#L2567

**Comment by fcoufour** [comment]
Hi @xet7,

If I understand the source code well, card creation through the API (https://github.com/wekan/wekan/blob/master/models/cards.js#L2511) is authorized for all users except commentOnly users with no Wekan admin rights and card edition (https://github.com/wekan/wekan/blob/master/models/cards.js#L2619) is authorized for Wekan admin users only.
Am I right? Is this a design choice to differ from the application rights?

NB: Moving cards specifically does not seem to work as expected (https://wekan.github.io/api/v5.14/#edit_card). Can you check this ?

`curl -X PUT <WEKAN_URL>/api/boards/<BOARD_ID>/lists/<OLD_LIST_ID>/cards/<CARD_ID> -H 'Content-Type: application/json' -H 'Authorization: Bearer <USER_TOKEN>' -d '{"listId":"<NEW_LIST_ID>"}'`

gives `500 Internal Server Error`

Thanks for your great work on Wekan by the way.

**Comment by xet7** [comment]
@fcoufour 

Thanks :)

Does this check is user admin?
https://github.com/wekan/wekan/blob/master/models/cards.js#L2518

Current application rights are combination of what I and other Wekan contributors have implemented.

Current API example code is at:
https://github.com/wekan/wekan/blob/master/api.py

It would be nice if someone could add pull request to that file, to add commands remaining of API.
Unfortunately I don't yet have time to do it, I have to look at my roadmap to see
could I somehow get those done that are near deadlines.

**Comment by fcoufour** [comment]
Thanks for your reply.

My main question was whether my understanding of the code relative to API user rights was correct and whether the implementation was in line with what is functionally desired for the project.

The issue with moving cards is the same with admin or normal user (not sure if this answers your question or not).

**Comment by fredlang7680** [comment]
Hi @xet7 ,
I use Wekan V5.90 with offline installation.
I have created a GANTT in Angular, and I use the Wekan API. All works well, I can login the user, load boards, lists, etc... 
But when I want to update the dates of a card via the PUT method, there is the message:

Access to XMLHttpRequest at 'http://xxxxx/api/boards/kzuQt6Q9npnwmbYdy/lists/g3XGJkFMKfgQb7ckC/cards/63n5cxXFEu9b2nC6p' from origin 'http://xxxxx' has been blocked by CORS policy: Method PUT is not allowed by Access-Control-Allow-Methods in preflight response

My config file:
SET CORS=*
SET CORS_ALLOW_HEADERS=Authorization,Content-Type
SET CORS_EXPOSE_HEADERS=*

but nothing for Access-Control-Allow-Methods ?

Here is my angular code:

**Service :**
----------------------------------------------------------
  Change_Date_Debut_Carte(boardid:string, listid:string, cardid:string,datedebut:Date): Observable<any> {  
    const httpOptions = {
      headers: new HttpHeaders({
        Accept:  'application/json',
        'Access-Control-Allow-Headers':'Content-Type, authorization',
        "Content-Type": "multipart/form-data"
      })
    };
      return this.http.put<any>(this.url + '/api/boards/' + boardid + '/lists/' + listid + '/cards/' + cardid, {startAt: datedebut}, httpOptions);
  }  
----------------------------------------------------------

**Interceptor :**
----------------------------------------------------------
intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    const currentUser = this.AuthenticationService.currentUserValue;
        const isLoggedIn = currentUser && currentUser.token;
        const isApiUrl = request.url.startsWith(environment.url);
       
        if (isLoggedIn && isApiUrl) {
          request = request.clone({
            setHeaders: {
                authorization: "Bearer " + currentUser.token
              },
        });
        }
    return next.handle(request);
  }
----------------------------------------------------------

My code is OK, because when I activate a CORS plugin in Chrome, it works well, the dates are changed. 
Sorry for my poor english,

Thanks a lot!

**Comment by xet7** [comment]
@fredlang7680 

Maybe you could send pull request to add more here?

https://github.com/wekan/wekan/blob/master/server/cors.js

**Comment by fredlang7680** [comment]
Thak you very much @xet7! Your product is very good!
I add this in bundle\programs\server\app\app.js:

  if (process.env.CORS_Access_Control_Allow_Methods) {
    WebApp.rawConnectHandlers.use(function (req, res, next) {
      res.setHeader('Access-Control-Allow-Methods', process.env.CORS_Access_Control_Allow_Methods);
      return next();
    });
  }

And this in start-wekan.bat
SET CORS_Access_Control_Allow_Methods=*

cheers

**Comment by fredlang7680** [comment]
only one question again, when we change the dates with API, there is no activity in wekan? When we change in wekan, an activity is created, but not with API. How can I do this? THKS

![image](https://user-images.githubusercontent.com/56043103/148365459-afe7a61c-6ec2-43ae-b6a7-cfb590f248a4.png)

**Comment by xet7** [comment]
This feature already exists at cards.js:1038 (newBoardId/newListId)
